nexus/src/app/instance.rs

```diff
@@ -1357,6 +1384,127 @@ impl super::Nexus {
         Ok(())
     }
 
+    /// Compares the existing Propolis record in CRDB (`db_instance`) with a new
+    /// instance runtime state supplied from a sled agent, determines whether
+    /// any Propolis VMMs will be left unused as a result of this transition,
+    /// and releases the resources that were reserved to them.
+    async fn reclaim_unused_propolis_resources(
+        &self,
+        opctx: &OpContext,
+        new_runtime: &nexus::InstanceRuntimeState,
+        db_instance: &nexus_db_model::Instance,
+    ) -> Result<(), Error> {
+        let log = &self.log;
+        let instance_id = db_instance.id();
+
+        // Handle changes to the instance's active Propolis ID. The only
+        // expected change is from the current ID to a migration destination ID.
+        let old_propolis_id = db_instance.runtime().propolis_id;
+        if new_runtime.propolis_id != old_propolis_id {
+            match db_instance.runtime().dst_propolis_id {
+                Some(dst_id) if dst_id != new_runtime.propolis_id => {
+                    warn!(log,
+                          "active Propolis did not change to migration target";
```

Comment:
Should this actually fail instead of just warn? An instance suddenly being served by a new separate propolis in the midst of a migration seems like a bug.

Reply:
I'll need to think about this one a little more. I agree that if we get to this path, we've definitely violated an invariant somewhere. I'm unsure what to do about it from this path, though. In theory, we can talk to any of the sleds involved from this path:

- this routine can return a response to this sled agent that tells it to do something specifically
- it can also use the `sled_id` and `dst_sled_id` from the DB to contact either of the other involved sled agents

We could maybe use that to try to get all the instances to stop and then report the instance is Failed...? I'll need to mull it over (and I'm very much open to suggestions here :)).

(The main things I want to avoid are (a) silently going on like nothing's wrong, and (b) panicking--because this might mean that something unexpected is in CRDB, I want to avoid having the sled that's making this call go to each Nexus in turn and perform an operation that will deterministically cause it to panic.)